### ISSUE TITLE
Robolectric native graphics times out with captureToImage in Compose

### ISSUE THREAD

**Issue by oas004** [robolectric/robolectric#8071]
Firstly, thanks for a supercool library! 👏 

### Description

Testing out the native graphics from the [Robolectric 4.10 Alpha 1](https://github.com/robolectric/robolectric/releases/tag/robolectric-4.10-alpha-1) release.  I tried to do simple screenshot testing in Jetpack Compose by using the `composeTestRule.onRoot().captureToImage()` to capture the bitmap of the composable. This works in the AndroidTest folder with emulator and not using robolectric, but for robolectric it times out after 2000ms with

```
D/HardwareRendererCompat: Failed to reflectively call HardwareRenderer#isDrawingEnabled. It probably doesn't exist on this platform. Returning true.

androidx.compose.ui.test.ComposeTimeoutException: Condition still not satisfied after 2000 ms
```
Should this work with the native graphics feature of this release? :) 

### Steps to Reproduce

Use robolectric version 4.10-alpha1. Then add a test case like this in your test folder

```kt
    @get:Rule
    val composeTestRule = createComposeRule()

    @get:Rule
    val testName = TestName()


    @GraphicsMode(GraphicsMode.Mode.NATIVE)
    @Test
    fun testNativeGraphicsApproach1() {
        composeTestRule.apply {
            setContent {
                Greeting("Hello!")
            }
        }
        // This would be ideal, but it does not work for some reason.
        composeTestRule.onRoot().captureToImage()
        composeTestRule.onRoot().printToLog("SCREEN NODE TREE")
    }

```

Running this will reproduce the timeout issue. This works in the AndroidTest folder with an emulator.

### Robolectric & Android Version
4.10-alpha01

### Link to a public git repo demonstrating the problem:

Here is the repo link: https://github.com/oas004/NativeGraphicsSample

Here is the link to the test case in question:
https://github.com/oas004/NativeGraphicsSample/blob/main/app/src/test/java/com/example/nativegraphicssample/NativeGraphicsTest.kt

**Comment by yschimke** [comment]
Looks like general fixes are being done inside compose, not sure if related.

https://cs.android.com/androidx/platform/frameworks/support/+/b5d1500e6cfd59dbd801ac77433ce7555fdf1669

**Comment by oas004** [comment]
Good idea! I tried updating the compose version to 1.4.0-rc01 and compiler version 1.4.3 which I think are the latest versions? :) It looked like it from (https://developer.android.com/jetpack/androidx/releases/compose) at least. It still times out in the tests. Updated the sample project with this as well now.

**Comment by utzcoz** [comment]
@oas004 Thanks for reporting this issue. I can confirm that commits that @yschimke mentioned were merged into 1.4.0-rc01. After some local debugging, looks like `ViewRootImpl#mSurface` is invalid in Robolectric environment, and `ViewRootImpl#draw` return before dispatching `ViewTreeObserver` callback to notify `WindowCapture.android.kt#View.forceRedraw`. So `WindowCapture.android.kt#View.forceRedraw` throws a timeout exception to you. @hoisie Hope this information can help you do deeper investigation.

**Comment by hoisie** [comment]
Thank you all for giving native graphics a try, and thanks for the repro project :)

Is this the implementation of captureToImage?
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/test-utils/src/androidMain/kotlin/androidx/compose/testutils/ViewCapture.android.kt;l=47

Just taking a cursory look at that logic, it seems problematic a Robolectric environment:
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/test-utils/src/androidMain/kotlin/androidx/compose/testutils/ViewCapture.android.kt;l=66-92

There is a task being posted to the main looper, and the test thread is blocking until that task completes. This type of pattern is generally not compatible with a Robolectric environment, since the test thread is the same as the UI thread.

I haven't run the test under a debugger, this is just my first impression.

**Comment by utzcoz** [comment]
@hoisie My quick Android Studio debugging tells me that it might be related to SurfaceControl's native shadow. But I don't have time to continue it today because too late now.

**Comment by oas004** [comment]
Might be wrong here, but isn't this the implementation? :)  https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-test/src/androidMain/kotlin/androidx/compose/ui/test/AndroidImageHelpers.android.kt?q=SemanticsNodeInteraction.captureToImage()

I was trying to use the same as in the Compose samples here https://github.com/android/compose-samples/blob/e6994123804b976083fa937d3f5bf926da4facc5/Rally/app/src/androidTest/java/com/example/compose/rally/ScreenshotComparator.kt

**Comment by hoisie** [comment]
cc @brettchabot @JuliaSullivanGoogle 

[View.forceRedraw](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-test/src/androidMain/kotlin/androidx/compose/ui/test/android/WindowCapture.android.kt;l=68-94) follows the same pattern: posting a task to the UI thread and blocking the current thread waiting to complete. This is not supported in Robolectric.

The stuff inside `handler.post(...)` does not executed until after the test times out and the activity is torn down.

I am not sure if Compose's `captureToImage` is supported in Robolectric. @JuliaSullivanGoogle has done Compose screenshots in RNG and may have better context.

**Comment by JuliaSullivanGoogle** [comment]
captureToImage does not work in Robolectric because it tries to invalidate the hierarchy and force a redraw, which doesn't work in Robolectric. I made a version internally that effectively does the same thing, but avoids redrawing since it isn't needed in Robolectric. ShadowNativePixelCopy works now, which is what the original implementation was using, so just using that in a kotlin extension function was all that was needed.

**Comment by devDebajo** [comment]
@JuliaSullivanGoogle, as I understand it, you succeded to run captureToImage in Robolectric. Could you show a code example?

**Comment by yschimke** [comment]
@devDebajo you should get the proper code from @JuliaSullivanGoogle but I'm working around it temporarily with

```
// store the view inside setContent
view = LocalView.current
```

Then

```
                    Bitmap.createBitmap(view.width, view.height, Bitmap.Config.ARGB_8888).apply {
                        view.draw(Canvas(this))
                    }
```

**Comment by utzcoz** [comment]
> captureToImage does not work in Robolectric because it tries to invalidate the hierarchy and force a redraw, which doesn't work in Robolectric. 

@JuliaSullivanGoogle Is it possible to support it? From my previous quick debugging, looks like `SurfaceControl` in `ViewRootImpl` is not initialized and processed correctly. At least `ViewRootImpl#redraw` can be called and run.

**Comment by devDebajo** [comment]
@yschimke I do not understand anything, but in my case any Bitmaps are empty, which rendered in Robolectric tests

**Comment by yschimke** [comment]
@devDebajo yep, I'm out of my depth.  Just trying to offer a quick way forward.

**Comment by devDebajo** [comment]
@yschimke does Bitmap and Canvas working in Robolectric tests?
I tried to create red bitmap inside test:
```
val bitmap = Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888)
val canvas = Canvas(bitmap)
canvas.drawColor(android.graphics.Color.RED)
```
And i've got empty Bitmap

**Comment by devDebajo** [comment]
Sorry for my carelessness. Native graphics introduced in 4.10-alpha-1. Red square successfully rendered on this version)

**Comment by yschimke** [comment]
I can't use `ComposeContentTestRule.awaitIdle()` in robolectric tests.  Is that the same issue?

**Comment by hoisie** [comment]
cc @jfresen from Compose

**Comment by anthony-fresneau-kiplin** [comment]
> captureToImage does not work in Robolectric because it tries to invalidate the hierarchy and force a redraw, which doesn't work in Robolectric. I made a version internally that effectively does the same thing, but avoids redrawing since it isn't needed in Robolectric. ShadowNativePixelCopy works now, which is what the original implementation was using, so just using that in a kotlin extension function was all that was needed.

@JuliaSullivanGoogle Can we have access to ShadowNativePixelCopy ?

**Comment by eygraber** [comment]
`ShadowPixelCopy` is available in 4.11-beta-2 which allows you to do what @JuliaSullivanGoogle suggested. Make sure you annotate your test / test class with:

```kotlin
@GraphicsMode(GraphicsMode.Mode.NATIVE)
@Config(shadows = [ShadowPixelCopy::class], minSdk = 26)
```

```kotlin
@OptIn(ExperimentalTestApi::class)
@RequiresApi(Build.VERSION_CODES.O)
private fun SemanticsNodeInteraction.captureToImage(): ImageBitmap {
  val node = fetchSemanticsNode("Failed to capture a node to bitmap.")
  // Validate we are in popup
  val popupParentMaybe = node.findClosestParentNode(includeSelf = true) {
    it.config.contains(SemanticsProperties.IsPopup)
  }
  if (popupParentMaybe != null) {
    return processMultiWindowScreenshot(node)
  }

  val view = (node.root as ViewRootForTest).view

  // If we are in dialog use its window to capture the bitmap
  val dialogParentNodeMaybe = node.findClosestParentNode(includeSelf = true) {
    it.config.contains(SemanticsProperties.IsDialog)
  }
  var dialogWindow: Window? = null
  if (dialogParentNodeMaybe != null) {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
      // TODO(b/163023027)
      throw IllegalArgumentException("Cannot currently capture dialogs on API lower than 28!")
    }

    dialogWindow = findDialogWindowProviderInParent(view)?.window
      ?: throw IllegalArgumentException(
        "Could not find a dialog window provider to capture its bitmap"
      )
  }

  val windowToUse = dialogWindow ?: view.context.getActivityWindow()

  val nodeBounds = node.boundsInRoot
  val nodeBoundsRect = Rect(
    nodeBounds.left.roundToInt(),
    nodeBounds.top.roundToInt(),
    nodeBounds.right.roundToInt(),
    nodeBounds.bottom.roundToInt()
  )

  val locationInWindow = intArrayOf(0, 0)
  view.getLocationInWindow(locationInWindow)
  val x = locationInWindow[0]
  val y = locationInWindow[1]

  // Now these are bounds in window
  nodeBoundsRect.offset(x, y)

  return windowToUse.captureRegionToImage(nodeBoundsRect)
}

@RequiresApi(Build.VERSION_CODES.O)
private fun SemanticsNode.findClosestParentNode(
  includeSelf: Boolean = false,
  selector: (SemanticsNode) -> Boolean
): SemanticsNode? {
  var currentParent = if (includeSelf) this else parent
  while (currentParent != null) {
    if (selector(currentParent)) {
      return currentParent
    } else {
      currentParent = currentParent.parent
    }
  }

  return null
}

@ExperimentalTestApi
@RequiresApi(Build.VERSION_CODES.O)
private fun processMultiWindowScreenshot(
  node: SemanticsNode
): ImageBitmap {
  val nodePositionInScreen = findNodePosition(node)
  val nodeBoundsInRoot = node.boundsInRoot

  val combinedBitmap = InstrumentationRegistry.getInstrumentation().uiAutomation.takeScreenshot()

  val finalBitmap = Bitmap.createBitmap(
    combinedBitmap,
    (nodePositionInScreen.x + nodeBoundsInRoot.left).roundToInt(),
    (nodePositionInScreen.y + nodeBoundsInRoot.top).roundToInt(),
    nodeBoundsInRoot.width.roundToInt(),
    nodeBoundsInRoot.height.roundToInt()
  )
  return finalBitmap.asImageBitmap()
}

private fun findNodePosition(
  node: SemanticsNode
): Offset {
  val view = (node.root as ViewRootForTest).view
  val locationOnScreen = intArrayOf(0, 0)
  view.getLocationOnScreen(locationOnScreen)
  val x = locationOnScreen[0]
  val y = locationOnScreen[1]

  return Offset(x.toFloat(), y.toFloat())
}

internal fun findDialogWindowProviderInParent(view: View): DialogWindowProvider? {
  if (view is DialogWindowProvider) {
    return view
  }
  val parent = view.parent ?: return null
  if (parent is View) {
    return findDialogWindowProviderInParent(parent)
  }
  return null
}

private fun Context.getActivityWindow(): Window {
  fun Context.getActivity(): Activity {
    return when (this) {
      is Activity -> this
      is ContextWrapper -> this.baseContext.getActivity()
      else -> throw IllegalStateException(
        "Context is not an Activity context, but a ${javaClass.simpleName} context. " +
          "An Activity context is required to get a Window instance"
      )
    }
  }
  return getActivity().window
}

@RequiresApi(Build.VERSION_CODES.O)
private fun Window.captureRegionToImage(
  boundsInWindow: Rect,
): ImageBitmap {
  // Turn on hardware rendering, if necessary
  return withDrawingEnabled {
    // Then we generate the bitmap
    generateBitmap(boundsInWindow).asImageBitmap()
  }
}

private fun <R> withDrawingEnabled(block: () -> R): R {
  val wasDrawingEnabled = HardwareRendererCompat.isDrawingEnabled()
  try {
    if (!wasDrawingEnabled) {
      HardwareRendererCompat.setDrawingEnabled(true)
    }
    return block.invoke()
  } finally {
    if (!wasDrawingEnabled) {
      HardwareRendererCompat.setDrawingEnabled(false)
    }
  }
}

@RequiresApi(Build.VERSION_CODES.O)
private fun Window.generateBitmap(boundsInWindow: Rect): Bitmap {
  val destBitmap =
    Bitmap.createBitmap(
      boundsInWindow.width(),
      boundsInWindow.height(),
      Bitmap.Config.ARGB_8888
    )
  generateBitmapFromPixelCopy(boundsInWindow, destBitmap)
  return destBitmap
}

@RequiresApi(Build.VERSION_CODES.O)
private object PixelCopyHelper {
  @DoNotInline
  fun request(
    source: Window,
    srcRect: Rect?,
    dest: Bitmap,
    listener: PixelCopy.OnPixelCopyFinishedListener,
    listenerThread: Handler
  ) {
    PixelCopy.request(source, srcRect, dest, listener, listenerThread)
  }
}

@RequiresApi(Build.VERSION_CODES.O)
private fun Window.generateBitmapFromPixelCopy(boundsInWindow: Rect, destBitmap: Bitmap) {
  val latch = CountDownLatch(1)
  var copyResult = 0
  val onCopyFinished = PixelCopy.OnPixelCopyFinishedListener { result ->
    copyResult = result
    latch.countDown()
  }
  PixelCopyHelper.request(
    this,
    boundsInWindow,
    destBitmap,
    onCopyFinished,
    Handler(Looper.getMainLooper())
  )

  if (!latch.await(1, TimeUnit.SECONDS)) {
    throw AssertionError("Failed waiting for PixelCopy!")
  }
  if (copyResult != PixelCopy.SUCCESS) {
    throw AssertionError("PixelCopy failed!")
  }
}
```

**Comment by oas004** [comment]
I tried to update it in the sample I made here (https://github.com/oas004/NativeGraphicsSample/pull/1) but with `captureToImage` it still times out but with a different error. Is there a method on `ShadowPixelCopy` that I was supposed to use instead? Maybe I did something wrong while upgrading.

**Comment by oas004** [comment]
I just got that the test is not giving any result within 2000ms, but I'm not sure I quite understood how I was supposed to test it.   I split it into different commits in the PR if you wanna check it out, but none of these approaches worked for me at least 🤔

**Comment by eygraber** [comment]
@oas004 it looks like you might still be using the library `captureToImage` because of this import `import androidx.compose.ui.test.*`

**Comment by oas004** [comment]
Aha! Thanks! Changing that made the test pass at least :) Sorry about that

**Comment by oas004** [comment]
@utzcoz Should we close this as completed now that the fix from @eygraber worked? :)

**Comment by yschimke** [comment]
I was hoping it would make the built in one work. Is that not feasible?

**Comment by eygraber** [comment]
I don't think it'll work because as @JuliaSullivanGoogle said

> captureToImage does not work in Robolectric because it tries to invalidate the hierarchy and force a redraw, which doesn't work in Robolectric.

All my workaround did was remove the call to force a redraw.

**Comment by eygraber** [comment]
Any updates on this?

**Comment by eygraber** [comment]
So is this going to be addressed in Robolectric or Compose (or both)? [My workaround](https://github.com/robolectric/robolectric/issues/8071#issuecomment-1774365222) still works, but it is a copy of 200+ lines of code from 2 years ago. 

It would be nice if this was officially fixed.

**Comment by realdadfish** [comment]
The workaround apparently stopped working with recent 4.17 / 4.18 snapshots and SDK level 37. The captured image is all black. Investigating this.

**Comment by realdadfish** [comment]
Ok, guess this is reported through #11392 already.
